File: trunk/COM/Samples/Ruby/Convert2PDF.rb
# Convert2PDF.rb script
# Part of PDFCreator
# License: GPL
# Homepage: http://www.sf.net/projects/pdfcreator
# Version: 1.0.0.0
# Date: April, 4. 2007
# Author: Frank Heindörfer
# Comments: This script convert a printable file in a pdf-file using 
#           the com interface of PDFCreator.

require 'win32ole'

SleepTime = 30
$readyState = 0

pdfcreator = WIN32OLE.new('PDFCreator.clsPDFCreator')
event = WIN32OLE_EVENT.new(pdfcreator)
event.on_event('eReady') do
 puts 'Ready'
 $readyState = 1
end
event.on_event('eError') do
 puts 'An error is occured!\nError [', 
  pdfcreator.cErrorDetail('Number'), ']: ',
  pdfcreator.cErrorDetail('Description')
 pdfcreator.cClose()
 pdfcreator = nil
 Process.exit
end

if ARGV.length == 0 then
 puts 'Syntax:\tConvert2PDF.rb <Filename>!'
 pdfcreator.cClose()
 pdfcreator = nil
 Process.exit
end

pdfcreator.cStart('/NoProcessingAtStartup')
pdfcreator.setproperty('cOption', 'UseAutosave', 1)
pdfcreator.setproperty('cOption', 'UseAutosaveDirectory', 1)
pdfcreator.setproperty('cOption', 'AutosaveFormat', 0) # 0 = PDF
DefaultPrinter = pdfcreator.cDefaultprinter
pdfcreator.setproperty('cDefaultprinter', 'PDFCreator')
pdfcreator.cClearCache()
pdfcreator.setproperty('cPrinterStop', false)


0.upto(ARGV.length - 1) do |i|
 ifname =  ARGV[i]
 if File.dirname(ifname) == '.'
  ifname = Dir.getwd.gsub(/\//,'\\') + '\\' + ifname
 end
 if !FileTest.exist?(ARGV[i]) then
  print 'Can''t find the file: ', ifname
  break
 end
 if !pdfcreator.cIsPrintable(ifname) then
  print 'Converting: ', ifname, '\n',
   ' An error is occured: File is not printable!'
  break
 end
 $readyState = 0

 pdfcreator.setproperty('cOption', 'AutosaveDirectory', File.dirname(ifname))
 pdfcreator.setproperty('cOption', 'AutosaveFilename', File.basename(ifname, File.extname(ifname)))
 pdfcreator.cPrintfile(ifname)

 1.upto SleepTime do
  pdfcreator.cOption('UseAutosave') # dummy command, otherwise the com server can't send an event to the script
  break if $readyState != 0
  sleep 1
 end
 if $readyState == 0 then
  print 'Converting: ', ifname, '\n',
   ' An error is occured: Time is up!'
  break 
 end
end

pdfcreator.setproperty('cDefaultprinter', DefaultPrinter)
pdfcreator.cClearCache()
pdfcreator.cClose()
pdfcreator = nil
Process.exit 

File: tags/Version 1.4.2/COM/Samples/Python/Convert2PDF.py
# -*- coding: ISO-8859-1 -*-
# Convert2PDF.py script
# Part of PDFCreator
# License: GPL
# Homepage: http://www.pdfforge.org/products/pdfcreator
# python 2.5, pywin32 2.5.1
# Version: 1.0.0.0
# Date: December, 24. 2007
# Author: Frank Heindörfer
# Comments: This script convert a printable file in a pdf-file using 
#           the com interface of PDFCreator.

import win32com.client as com
from win32print import GetDefaultPrinter, SetDefaultPrinter 
import os
import sys

from time import sleep
from pythoncom import PumpWaitingMessages, PumpMessages

sleepTime = 1  # in seconds
maxTime = 10   # in seconds
ReadyState = 0

class clsPDFCreatorEvents(object):
	def OneReady(self):
		print "Ready."
		global ReadyState
		ReadyState = 1
	def OneError(self):
		ReadyState = 0
		global PDFCreator
		print "An error is occured!\r\n Error [" + \
		str(PDFCreator.cErrorDetail("Number")) + "]: " + \
		PDFCreator.cErrorDetail("Description")
		errNum = PDFCreator.cErrorDetail("Number")
		PDFCreator = None
		os._exit(errNum)

if len(sys.argv) <= 1:
	print "Syntax: " + os.path.basename(sys.argv[0]) + " <Filename>\r\n\tor use ""Drag and Drop""!"
	os._exit(-255)

PDFCreator = com.Dispatch('PDFCreator.clsPDFCreator')
PDFCreatorEvents = com.DispatchWithEvents(PDFCreator, clsPDFCreatorEvents)
ProgramIsRunning = PDFCreator.cProgramIsRunning
PDFCreator.cStart("/NoProcessingAtStartup", 1)

options = PDFCreator.cOptions
oldOptions = PDFCreator.cOptions
options.UseAutosave = 1
options.UseAutosaveDirectory = 1
options.AutosaveFormat = 0                            # 0 = PDF
CurrentPrinter = GetDefaultPrinter()
SetDefaultPrinter('PDFCreator')
PDFCreator.cClearCache
PDFCreator.cPrinterStop = 0

for i in xrange(1, len(sys.argv)):
	ifname = str(sys.argv[i])
	if not os.path.exists(ifname):
		print "Can't find the file: " + ifname
		exit()
	if not PDFCreator.cIsPrintable(ifname):
		print "Converting: " + ifname + "\r\n\r\nAn error is occured: File is not printable!"
		exit()
	ReadyState = 0
	dirname = os.path.dirname(ifname)
	if dirname == "":
		dirname = os.path.dirname(sys.argv[0])
	options.AutosaveDirectory = dirname
	options.AutosaveFilename = os.path.splitext(os.path.basename(ifname))[0]
	PDFCreator.cSaveOptions(options)
	
	PDFCreator.cPrintFile(ifname)
	
	c = 0
	while (ReadyState == 0) and (c < (maxTime / sleepTime)):
		c = c + 1
		sleep(sleepTime)
		PumpWaitingMessages()
	if ReadyState == 0:
		print "Converting: " + ifname + "\r\n\r\nAn error is occured: Time is up!"
		exit()
		
PDFCreator.cOptions = options
PDFCreator.cSaveOptions(oldOptions)
SetDefaultPrinter(CurrentPrinter)
PDFCreator.cClose
PDFCreator = None
print "Done."